Fct convert table to tree

susanm74 opened the pull request:
1. fct table converted to tree... and my DEEEEEPEST apologies for not putting these into separate branches...
2. cleaned up dashboard/dsr/fct views and templates and converted single quotes to double quotes, 
3. fixed dashboard 'level up' link to hide when current program/location parent == null, 
4. fixed link param filter location types param when updating dsr category, fct target, or top-level facility types, 
5. fixed filter tabs program param when switching between reports,
   6.updated tests!!!

susanm74 commented:
Whoever bears through this, I will +1 you a BEER, DUDE!

fterrier commented on test/integration/org/chai/kevin/IntegrationTests.groovy:
```diff
@@ -172,6 +211,57 @@ abstract class IntegrationTests extends IntegrationSpec {
 		return entity
 	}
 	
+	static def newDashboardProgram(def code, def program) {
```
Any reasons for this being moved into IntegrationTests? I guess they should be in the report-specfic Dashboard/DSR/FCTIntegrationTests files ...

fterrier commented on grails-app/taglib/org/chai/kevin/FilterTagLib.groovy:
```diff
@@ -125,7 +151,53 @@ class FilterTagLib {
+		}				
+		
+		def programTargetClass = null
+		switch(tab){
```
ooops, do we really need to introduce specific cases for each report type ?

susanm74 replied:
Ex. If the dashboard program id doesn't exist for dsr/fcts, the tab link should be updated before its displayed, so that the link contains the program id that _will_ display when you nav to the dsr/fct page, which is needed to correctly set the top-level program filter tree, report table headers, etc. When the program id contains the erroneous dashboard program id, those things are incorrectly set (displaying "Blah" as the selected "currentProgram" when the program tree doesn't contain "Blah", and displaying "Blah" as the current program in the report table headers, and the page displaying no data, all of which are unhelpful to the user, and when I was testing, looked sloppy to not reset to the root program).

On the other hand, maybe we could grab the program tree in the controller instead of createLinkByTabs, and compare against that and update the model before displaying the view (or redirect), its essentially what the createLinkByTabs method does, which would achieve the same thing, but without also fixing the tab links.

fterrier commented on grails-app/views/templates/_reportTableHeader.gsp:
```diff
@@ -37,17 +37,17 @@
\ No newline at end of file
+<g:if test="${controllerName == 'dashboard'}">
+	<% def levelUpLinkParams = new HashMap(linkParams) %>
+	<g:if test="${table == 'program' && currentProgram.parent != null}">
```
is this distinction needed ? these two bodies look exactly the same

susanm74 replied:
I think so, because sometimes the program will be set to the root, and the location will not, and vice versa, so one table will need to display the link, but the other will not.

fterrier commented on grails-app/views/templates/_topLevelReportTabs.gsp:
```diff
@@ -1,19 +1,15 @@
-	
 	<li><a ${controllerName == 'dashboard' ? 'class="selected"':''}
-		href="${createLink(controller:'dashboard', action:actionName, params:linkParams)}">Performance</a></li>
+		href="${createLinkByTab(controller:'dashboard', action:actionName, params:linkParams)}">
```
Do we need to modify the link for that? Maybe we could generate the link using the normal createLink method but have a redirect happen in the controller in case the respective objective doesn't exist for the selected report type. This would also solve the issue of the method createLinkByTab contain a switch/case for each report type.

fterrier commented on grails-app/views/templates/_reportTableHeader.gsp:
```diff
@@ -21,13 +21,13 @@
 	</g:elseif>
 		<g:if test="${controllerName == 'fct'}">
-		<g:if test="${table == 'location'}">
+		<g:if test="${table == 'program'}">
```
Is this needed ? Or can we use CSS classes instead of an if/else ?

susanm74 committed:
1. added dsr controller int tests, 2. moved param value 'to list' code back into link param filter, 3. moved report integration test helper methods into separate files, 4. separated integration test program tree setup from report-specific program tree setup

susanm74 committed:
fixed link param filter param value bug and updated tests

susanm74 committed:
Merge branch 'master' into fct-convert-table-to-tree

susanm74 commented:
UPDATE
1. added dsr controller int tests, 
2. moved param value 'to list' code back into link param filter, 
3. moved report integration test helper methods into separate files, 
4. separated integration test program tree setup from report-specific program tree setup

fterrier commented:
ship it dude
